runtime/src/main/java/org/corfudb/runtime/collections/CorfuStore.java

```diff
@@ -218,19 +224,61 @@ public TxnContext txn(@Nonnull final String namespace, IsolationLevel isolationL
     }
 
     /**
-     * Return the address of the latest updated made in this table.
+     * Return the address of the latest update made in this table.
+     *
+     * Note: we can't deliberately return the tail of the stream map
+     * as this entry might be a HOLE, we need to filter and return only that
+     * corresponding to the last DATA entry.
      *
      * @param namespace - namespace that this table belongs to.
      * @param tableName - table name of this table without the namespace prefixed in.
      * @return stream tail of this table.
      */
     public long getHighestSequence(@Nonnull final String namespace,
                                    @Nonnull final String tableName) {
-        return this.runtime.getSequencerView().query(
-                CorfuRuntime.getStreamID(
-                        TableRegistry.getFullyQualifiedTableName(namespace, tableName)
-                )
-        );
+
+        // Overall strategy:
+        // (1) Retrieve streams full address map from sequencer
+        // (2) Read entries in reverse to confirm it's a DATA type entry (for optimization
+        // read in batches). A batch size of 4 is guaranteed to suit one edge case case scenario: i.e.,
+        // consecutive checkpoints enforcing holes (3 rounds of CP accumulate before trim happens == 3 holes)
+
+        long startTime = System.nanoTime();
+        UUID streamId = CorfuRuntime.getStreamID(
+                TableRegistry.getFullyQualifiedTableName(namespace, tableName));
+        StreamAddressSpace streamAddressSpace = this.runtime.getSequencerView()
+                .getStreamAddressSpace(new StreamAddressRange(streamId, Address.MAX, Address.NON_ADDRESS));
+
+        int numBatches = 0;
+
+        if (streamAddressSpace.size() != 0) {
+            NavigableSet<Long> addresses = streamAddressSpace.copyAddressesToSet(Address.MAX).descendingSet();
+            Iterable<List<Long>> batches = Iterables.partition(addresses,
+                    runtime.getParameters().getHighestSequenceNumberBatchSize());
+
+            for (List<Long> batch : batches) {
+                numBatches++;
+                // TODO: We can optimize this such that actual data is not transferred back to the client,
+                //  we require an API that inspects the data at the remote and indicates whether its an actual DATA
+                //  entry or a HOLE, e.g.: Map<Long, DataType> entryTypes = runtime.getAddressSpaceView().getDataType(batch)
```

Comment:
Is it possible that we have an API to get stream's non-HOLE tail sequence number? It will have only one RPC

Reply:
It is, we discussed the same, just want to do this in phases, first phase we do the basic client only fix. Next phase either an rpc that returns the first non-hole address or an rpc that does not transfer data.